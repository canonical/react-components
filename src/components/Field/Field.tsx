import classNames from "classnames";
import React from "react";
import type { ReactNode } from "react";

import Label from "../Label";
import Col from "../Col";

import type { ClassName } from "types";

/**
 * The props for the Field component.
 */
export type Props = {
  /**
   * The content for caution validation.
   */
  caution?: ReactNode;
  /**
   * The content of the Field component.
   */
  children?: ReactNode;
  /**
   * Optional class(es) to pass to the Field wrapper div.
   */
  className?: ClassName;
  /**
   * The content for error validation.
   */
  error?: ReactNode;
  /**
   * The id of the input this Field component is controlling.
   */
  forId?: string;
  /**
   * Help text to show below the field.
   */
  help?: ReactNode;
  /**
   * Optional class(es) to pass to the help text element.
   */
  helpClassName?: ReactNode;
  /**
   * An id to give to the help element.
   */
  helpId?: string;
  /**
   * Whether the component is wrapping a select element.
   */
  isSelect?: boolean;
  /**
   * Whether the component is wrapping a checkbox or radio element.
   */
  isTickElement?: boolean;
  /**
   * The label for the field.
   */
  label?: ReactNode;
  /**
   * Optional class(es) to pass to the label component.
   */
  labelClassName?: string;
  /**
   * Whether the label should show before the input.
   */
  labelFirst?: boolean;
  /**
   * Whether the field is required.
   */
  required?: boolean;
  /**
   * Whether the form field should have a stacked appearance.
   */
  stacked?: boolean;
  /**
   * The content for success validation.
   */
  success?: ReactNode;
  /**
   * An id to give to the caution, error or success element.
   */
  validationId?: string;
};

const generateHelpText = ({
  help,
  helpId,
  helpClassName,
  isTickElement,
}: Pick<Props, "help" | "helpId" | "helpClassName" | "isTickElement">) =>
  help ? (
    <p
      className={classNames("p-form-help-text", helpClassName, {
        "is-tick-element": isTickElement,
      })}
      id={helpId}
    >
      {help}
    </p>
  ) : null;

const generateError = (
  error: Props["error"],
  caution: Props["caution"],
  success: Props["success"],
  validationId: Props["validationId"]
) => {
  if (!error && !caution && !success) {
    return null;
  }
  const messageType =
    (error && "Error") || (caution && "Caution") || (success && "Success");
  return (
    <p className="p-form-validation__message" id={validationId}>
      <strong>{messageType}:</strong> {error || caution || success}
    </p>
  );
};

const generateLabel = (
  forId: Props["forId"],
  required: Props["required"],
  label: Props["label"],
  labelClassName: Props["labelClassName"],
  stacked: Props["stacked"]
) => {
  if (!label) {
    return null;
  }
  const labelNode = (
    <Label className={labelClassName} forId={forId} required={required}>
      {label}
    </Label>
  );
  if (stacked) {
    return <Col size={4}>{labelNode}</Col>;
  }
  return labelNode;
};

const generateContent = ({
  isSelect,
  children,
  labelFirst,
  labelNode,
  help,
  helpClassName,
  error,
  caution,
  success,
  validationId,
  helpId,
  isTickElement,
}: Partial<Props> & {
  labelNode: JSX.Element | null;
  validationId: string;
  helpId: string;
}) => (
  <div className="p-form__control u-clearfix">
    {isSelect ? (
      <div className="p-form-validation__select-wrapper">{children}</div>
    ) : (
      children
    )}
    {!labelFirst && labelNode}
    {generateHelpText({
      helpId,
      help,
      helpClassName,
      isTickElement,
    })}
    {generateError(error, caution, success, validationId)}
  </div>
);

const Field = ({
  caution,
  children,
  className,
  error,
  forId,
  help,
  helpClassName,
  helpId,
  isSelect,
  isTickElement,
  label,
  labelClassName,
  labelFirst = true,
  required,
  stacked,
  success,
  validationId,
  ...props
}: Props): JSX.Element => {
  const labelNode = generateLabel(
    forId,
    required,
    label,
    labelClassName,
    stacked
  );

  const content = generateContent({
    isSelect,
    isTickElement,
    children,
    labelFirst,
    labelNode,
    help,
    helpClassName,
    error,
    caution,
    success,
    validationId,
    helpId,
  });
  return (
    <div
      className={classNames("p-form__group", "p-form-validation", className, {
        "is-error": error,
        "is-caution": caution,
        "is-success": success,
        row: stacked,
      })}
      {...props}
    >
      {labelFirst && labelNode}
      {stacked ? <Col size={8}>{content}</Col> : content}
    </div>
  );
};

export default Field;
