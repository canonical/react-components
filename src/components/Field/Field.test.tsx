import { render, screen } from "@testing-library/react";
import React from "react";

import Field from "./Field";

describe("Field ", () => {
  it("renders", () => {
    render(
      <Field forId="test-id" label="Test label" data-testid="field">
        Test content
      </Field>
    );
    expect(screen.getByTestId("field")).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    render(<Field className="extra-class" data-testid="field" />);
    const field = screen.getByTestId("field");
    expect(field).toHaveClass("p-form__group");
    expect(field).toHaveClass("extra-class");
  });

  it("can display a caution message", () => {
    render(
      <Field caution="Are you sure?" validationId="id-1" data-testid="field">
        <input aria-errormessage="id-1" aria-invalid="true" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveErrorMessage(
      "Caution: Are you sure?"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-caution");
  });

  it("can display a caution node", () => {
    render(
      <Field
        caution={<span>Are you sure?</span>}
        validationId="id-1"
        data-testid="field"
      >
        <input aria-errormessage="id-1" aria-invalid="true" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveErrorMessage(
      "Caution: Are you sure?"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-caution");
  });

  it("can display an error message", () => {
    render(
      <Field error="You can't do that" validationId="id-1" data-testid="field">
        <input aria-errormessage="id-1" aria-invalid="true" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveErrorMessage(
      "Error: You can't do that"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-error");
  });

  it("can display an error node", () => {
    render(
      <Field
        error={<span>You can't do that</span>}
        validationId="id-1"
        data-testid="field"
      >
        <input aria-errormessage="id-1" aria-invalid="true" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveErrorMessage(
      "Error: You can't do that"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-error");
  });

  it("can display a success message", () => {
    render(
      <Field success="You did it!" validationId="id-1" data-testid="field">
        <input aria-describedby="id-1" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveAccessibleDescription(
      "Success: You did it!"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-success");
  });

  it("can display a success node", () => {
    render(
      <Field
        success={<span>You did it!</span>}
        validationId="id-1"
        data-testid="field"
      >
        <input aria-describedby="id-1" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveAccessibleDescription(
      "Success: You did it!"
    );
    expect(screen.getByTestId("field")).toHaveClass("is-success");
  });

  it("can display a help message", () => {
    render(
      <Field help="This is how you do it" helpId="id-1">
        <input aria-describedby="id-1" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveAccessibleDescription(
      "This is how you do it"
    );
  });

  it("can display a help node", () => {
    render(
      <Field help={<span>This is how you do it</span>} helpId="id-1">
        <input aria-describedby="id-1" />
      </Field>
    );
    expect(screen.getByRole("textbox")).toHaveAccessibleDescription(
      "This is how you do it"
    );
  });

  it("can display the label before the input", () => {
    render(
      <Field
        labelFirst={true}
        label="Test label"
        labelClassName="label-node"
        data-testid="field"
      />
    );
    const field = screen.getByTestId("field");
    expect(field.childNodes[0]).toHaveClass("label-node");
    // The label should not be inside the control.
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-form__control .p-form__label")
    ).not.toBeInTheDocument();
  });

  it("can display the label after the input", () => {
    render(<Field labelFirst={false} label="Test label" data-testid="field" />);
    const field = screen.getByTestId("field");
    // The first element should not be the label node.
    expect(field.childNodes[0]).not.toHaveClass("label-node");
    // The label should be inside the control.
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-form__control .p-form__label")
    ).toBeInTheDocument();
  });

  it("can display a label node", () => {
    render(<Field label={<span>Test label</span>} />);
    expect(screen.getByText("Test label")).toBeInTheDocument();
  });

  it("can pass a class name to the label", () => {
    render(<Field label="Test label" labelClassName="custom-label" />);
    expect(screen.getByText("Test label")).toHaveClass("custom-label");
  });

  it("can be required", () => {
    render(<Field required={true} label="Test label" />);
    expect(screen.getByText("Test label")).toHaveClass("is-required");
  });

  it("can be stacked", () => {
    render(<Field stacked={true} label="Test label" data-testid="field" />);
    // The Label should be inside a col-4.
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".col-4 .p-form__label")
    ).toBeInTheDocument();
    // The control should be inside a col-8.
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".col-8 .p-form__control")
    ).toBeInTheDocument();
    expect(screen.getByTestId("field")).toHaveClass("row");
  });
});
