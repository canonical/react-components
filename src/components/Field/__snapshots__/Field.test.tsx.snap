// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field  renders 1`] = `
<div
  class="p-form__group p-form-validation"
  data-testid="field"
>
  <label
    class="p-form__label"
    for="test-id"
  >
    Test label
  </label>
  <div
    class="p-form__control u-clearfix"
  >
    Test content
  </div>
</div>
`;
